C8 api intermediate tutorial (#5636)
camunda/camunda-docs: docs/apis-tools/camunda-api-rest/intermediate-tutorial.md

```javascript
async function deployResources() {
  const accessToken = await getAccessToken(authorizationConfiguration);
}
```

4. Using your generated client credentials from [prerequisites](#prerequisites), capture your Zeebe API URL beneath your call for an access token by defining `camundaApiUrl`:

```javascript
const camundaApiUrl = process.env.CAMUNDA_REST_ADDRESS;
```

5. On the next line, script the API endpoint to list the existing roles:

```javascript
const url = `${camundaApiUrl}/deployments`;
```

6. We will now configure the variables representing the BPMN file and its form data. This may look different depending on which resources you choose to deploy, but reflects the block-scoped local variables and append method to insert a set of objects for the BPMN resource of this tutorial:

```javascript
const formData = new FormData();
// Read the BPMN file and add it to the form data
const bpmnFilePath = path.resolve("resources/calculate-sales-tax.bpmn");
const fileContent = fs.readFileSync(bpmnFilePath);
formData.append("resources", fileContent, {
  filename: "calculate-sales-tax.bpmn",
  contentType: "application/xml",

> });

Reviewer: non-blocking: there are a few bits of this code that would cause the upload to fail, without giving any sort of squiggly error in the editor: 

1. the name of the field on 93 must be `resources`, because the API requires it
2. The path to the file on 91 must be correct
3. The contentType on 95 must be `application/xml` (or at least, it should be, and I don't know what will happen if it isn't)

Is there anything we should add to specify these details?

:::

The existing process definition key and ID will now output. If you have an invalid API name or action name, or no arguments provided, or improper/insufficient credentials configured, an error message will output as outlined in the `cli.js` file.

## Create and start a process instance (POST)

To create and start a process instance based on the process instance key obtained in the request above, take the following steps:

1. Outline your function, similar to the steps above:

```javascript
async function createInstance([processDefinitionKey]) {
  const accessToken = await getAccessToken(authorizationConfiguration);
  const camundaApiUrl = process.env.CAMUNDA_REST_ADDRESS;
  const url = `${camundaApiUrl}/process-instances`;
}
```

2. Configure the API call:

```javascript
const options = {
  method: "POST",
  url,
  headers: {
    Accept: "application/json",
    Authorization: `Bearer ${accessToken}`,
  },
  data: {
    processDefinitionKey: processDefinitionKey,
    variables: {
      total: 90.0,

>     },

Reviewer: non-blocking: similar to a previous comment, the request would succeed even if the variable names were different, but the process instance itself wouldn't function as expected. Is there anything we should add to make it clear that the variable name is important?

  },
};
```

3. Process the results from the API call. For example:

```javascript
try {
  const response = await axios(options);
  const processInstance = response.data;

  console.log(`Process Instance Key: ${processInstance.processInstanceKey}`);
} catch (error) {
  console.error(error.message);
}
```

> 4. In your terminal, run `node cli.js processInstances create <key>`, where `<key>` is the process definition key.

Reviewer: non-blocking: add something about seeing the processInstanceKey in the output, and capturing it for the final method?

## Retrieve a process instance (GET)

To retrieve a process instance by the process instance key, take the following steps:

1. Outline your function, similar to the steps above:

```javascript
async function viewInstance([processInstanceKey]) {
  const accessToken = await getAccessToken(authorizationConfiguration);
  const camundaApiUrl = process.env.CAMUNDA_REST_ADDRESS;
  const url = `${camundaApiUrl}/process-instances/${processInstanceKey}`;
}
```

2. Configure the API call:

```javascript
const options = {
  method: "GET",
  url,
  headers: {
    Accept: "application/json",
    Authorization: `Bearer ${accessToken}`,
  },
};
```

3. Process the results from the API call. For example:

```javascript
try {
  const response = await axios(options);
  const results = response.data;

  console.log(
    `Process instance name: ${results.processDefinitionName}; State: ${results.state};`
  );
} catch (error) {
  console.error(error.message);
}
```

> 4. In your terminal, run `node cli.js processInstances view <key>`, where `<key>` is the process instance key.

Reviewer: non-blocking: add something about what they should expect to see in the output?

---
id: camunda-8-api-intermediate-tutorial
title: Intermediate tutorial
sidebar_label: Intermediate

> description: "Step through our intermediate Camunda 8 API tutorial to deploy resources, create and start a process instance, and view a process instance by its key."

Reviewer 2: ```suggestion
description: "Step through this intermediate Camunda 8 API tutorial to deploy resources, create and start a process instance, and view a process instance by its key."
```
Rewording suggestion

---

> 

Reviewer 2: Could add an intermediate badge, as per the Get Started guides?

In this tutorial, we'll step through examples to highlight the capabilities of the Camunda 8 API, such as deploying resources, creating and starting a process instance, and viewing a process instance by its key.

> 

Reviewer 2: Could add an "About this tutorial" section here, that includes who this is aimed at, link back to the beginner tutorial for example, and describe what is achieved at the end of the tutorial - like a learning statement at the beginning of a lesson (already covered in the intro sentence, but perhaps more detail).

> ## Prerequisites

Reviewer 2: Could make this section into a table layout to make it easier to see, as it is quite cluttered atm with a lot of information. For example, https://docs.camunda.io/docs/next/components/modeler/web-modeler/idp/idp-configuration/#prerequisites.

Author: Good tip. Adjusted with incoming commit 👍

- If you haven't done so already, [create a cluster](/guides/create-cluster.md).
- Upon cluster creation, [create your first client](/guides/setup-client-connection-credentials.md). Ensure you check the `Zeebe` client scope box.

> :::note

Reviewer 2: ```suggestion
:::caution
```
Is it worth making this a caution as this is important for them to capture first-time?

Make sure you keep the generated client credentials in a safe place. The **Client secret** will not be shown again. For your convenience, you can also download the client information to your computer.
:::

> - In this tutorial, we utilize a JavaScript-written [GitHub repository](https://github.com/camunda/camunda-api-tutorials) to write and run requests. Clone this repo before getting started.

Reviewer 2: ```suggestion
- In this tutorial, we use a JavaScript-written [GitHub repository](https://github.com/camunda/camunda-api-tutorials) to write and run requests. Clone this repo before getting started.
```
Rewording suggestion for plain English - I always trya nd avoid utilize where use can do the same job 😊

Author: Adjusted with incoming commit 👍

> - In this tutorial, we utilize a JavaScript-written [GitHub repository](https://github.com/camunda/camunda-api-tutorials) to write and run requests. Clone this repo before getting started.

Reviewer 2: Could front-load this so the most important bit (Clone this repo) is at the front of the sentence, and not missed.
For example: Clone the Camunda API Tutorials GitHub repository as we will use this to write and run requests during this tutorial (or something like this).

Author: Adjusted with the table 👍

- The first request we will run is a request to deploy one or more resources (such as processes, decision models, or forms). For the purposes of this tutorial, we have preconfigured a BPMN diagram and converted this into XML. This diagram, `calculate-sales-tax.bpmn`, can be found in the GitHub repository above within the `resources` folder. The BPMN diagram itself represents a process to calculate the total sales tax for a given purchase. You can take a closer look at this diagram by opening it in [Modeler](/components/modeler/about-modeler.md).
- Ensure you have [Node.js](https://nodejs.org/en/download) installed as this will be used for methods that can be called by the CLI (outlined later in this guide). Run `npm install` to ensure you have updated dependencies.

## Getting started

> - You need authentication to access the API endpoints. Find more information [here](./camunda-api-rest-authentication.md).

Reviewer 2: Could/should this be moved into the prerequisites? Also, perhaps change the link from a 'click here' text?

Author: Adjusted with incoming commit 👍

## Set up authentication

If you're interested in how we use a library to handle auth for our code, or to get started, examine the `auth.js` file in the GitHub repository. This file contains a function named `getAccessToken` which executes an OAuth 2.0 protocol to retrieve authentication credentials based on your client ID and client secret. Then, we return the actual token that can be passed as an authorization header in each request.

> To set up your credentials, create an `.env` file which will be protected by the `.gitignore` file. You will need to add the following:

Reviewer 2: ```suggestion
To set up your credentials, create an `.env` file which will be protected by the `.gitignore` file. Add the following environment variables:
```

- `CAMUNDA_CLIENT_ID`
- `CAMUNDA_CLIENT_SECRET`
- `CAMUNDA_REST_ADDRESS` (after creating a client and downloading the .env variables, this is reflected in the Console UI as `ZEEBE_REST_ADDRESS`)
- `CAMUNDA_TOKEN_AUDIENCE` (represented as `ZEEBE_TOKEN_AUDIENCE` in the Console UI), which is `zeebe.camunda.io` in a Camunda 8 SaaS environment. For example, your audience may be defined as `CAMUNDA_TOKEN_AUDIENCE=zeebe.camunda.io`.

These keys will be consumed by the `auth.js` file to execute the OAuth protocol, and should be saved when you generate your client credentials in [prerequisites](#prerequisites).

> Examine the existing `.env.example` file for an example of how your `.env` file should look upon completion. Do not place your credentials in the `.env.example` file, as this example file is not protected by the `.gitignore`.

Reviewer 2: ```suggestion
See the existing `.env.example` file for an example of how your `.env` file should look upon completion. Do not place your credentials in the `.env.example` file, as this example file is not protected by the `.gitignore`.
```

:::

7. Configure your POST request to the appropriate endpoint, including an authorization header based on the previously acquired `accessToken`:

```javascript
const options = {
  method: "POST",
  url,
  headers: {
    Accept: "application/json",
    Authorization: `Bearer ${accessToken}`,
    ...formData.getHeaders(),
  },
  data: formData,
};
```

8. Call the endpoint, process the results from the API call, and emit an error message from the server if necessary:

```javascript
try {
  const response = await axios(options);
  const deployedResources = response.data.deployments;

  // Emit deployed resources
  deployedResources.forEach((x) =>
    console.log(
      `Process Definition Key: ${x.processDefinition.processDefinitionKey}; Process Definition Id: ${x.processDefinition.processDefinitionId}`
    )
  );
} catch (error) {
  // Emit an error from the server.
  console.error(error.message);
}
```

9. In your terminal, run `node cli.js processInstances deploy`.

:::note

> This `deploy` command is connected to the `deployResources` function at the bottom of the `camunda-process-instances.js` file, and executed by the `cli.js` file. While we will work with roles in this tutorial, you may add additional arguments depending on the API calls you would like to make.

Reviewer 2: ```suggestion
This `deploy` command is connected to the `deployResources` function at the bottom of the `camunda-process-instances.js` file, and executed by the `cli.js` file. While we will work with roles in this tutorial, you may add additional arguments depending on the API requests you want to make.
```
Should we use request in place of call?

## Retrieve a process instance (GET)

To retrieve a process instance by the process instance key, take the following steps:

1. Outline your function, similar to the steps above:

```javascript
async function viewInstance([processInstanceKey]) {
  const accessToken = await getAccessToken(authorizationConfiguration);
  const camundaApiUrl = process.env.CAMUNDA_REST_ADDRESS;
  const url = `${camundaApiUrl}/process-instances/${processInstanceKey}`;
}
```

2. Configure the API call:

```javascript
const options = {
  method: "GET",
  url,
  headers: {
    Accept: "application/json",
    Authorization: `Bearer ${accessToken}`,
  },
};
```

3. Process the results from the API call. For example:

```javascript
try {
  const response = await axios(options);
  const results = response.data;

  console.log(
    `Process instance name: ${results.processDefinitionName}; State: ${results.state};`
  );
} catch (error) {
  console.error(error.message);
}
```

4. In your terminal, run `node cli.js processInstances view <key>`, where `<key>` is the process instance key. The `processDefinitionName` and `state` will then display in the output.

> ## If you get stuck

Reviewer 2: ```suggestion
## Troubleshooting
```
